### PR TITLE
Add spawn_future()

### PR DESCRIPTION
**Pull request by ispeters**
This diff adds `unifex::spawn_future(Sender, Scope)`.

### CODE REVIEW

File: test/spawn_future_test.cpp [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+  unifex::sync_wait(scope.join());
+
+  EXPECT_EQ(valueConstructs, valueDestructs);
+}
+
+TEST(spawn_future_test, spawn_future_of_never_sender_does_not_hang) {
+  unifex::v2::async_scope scope;
+  auto future = unifex::spawn_future(unifex::never_sender{}, scope);
+
+  unifex::sync_wait(unifex::when_all(
```
Comment (janondrusek) [inline test/spawn_future_test.cpp:185]: same as below: Is there a way to assert that cancellation took effect?

File: include/unifex/spawn_future.hpp [hunk 2]
```diff
@@ -0,0 +1,1029 @@
...
+  void abandon() noexcept {
+    // abandon() is only invoked from a stop callback that's registered when the
+    // future is awaited; we know for sure that the future has been started and
+    // will consume the spawned operation.  if abandonment fails, it means the
+    // stop request has lost the race to a natural completion and the future
+    // will produce whatever the operation produced.
+
+    // the operation must either be in the init state or one of the natural
+    // completion states; in the former case we want to mark it as abandoned and
+    // in the latter we'll just allow the future to complete naturally
```
[the remark this answers is not in the document]
Comment (ispeters) [reply include/unifex/spawn_future.hpp:159]: If the operation is in the `init` state then the `if`'s condition will be true; I've added an `else` that asserts that, when the state _isn't_ `ini`, it must be one of `value`, `error`, or `done`, as described in the comment.

File: test/spawn_future_test.cpp (hunk 1 above)
Comment (ispeters) [reply test/spawn_future_test.cpp:185]: `never_sender` will never complete unless it receives a stop request.  If cancellation doesn't take effect, the spawned `never_sender` operation will never complete, leaving the scope open forever, leaving the `scope.join()` (and thus the whole test case) blocked forever.

In other words, if this test finishes, cancellation took effect; conversely, if cancellation doesn't take effect, the test will hang forever.

Does that satisfy or would you prefer a more affirmative assertion?

File: test/spawn_future_test.cpp [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+      noop_receiver_with_scheduler) noexcept {
+    return unifex::inline_scheduler{};
+  }
+};
+
+TEST(spawn_future_test, discarding_connected_future_cancels_spawned_operation) {
+  unifex::v2::async_scope scope;
+  auto future = unifex::spawn_future(unifex::never_sender{}, scope);
+
+  unifex::sync_wait(unifex::when_all(
```
[the remark this answers is not in the document]
Comment (ispeters) [reply test/spawn_future_test.cpp:209]: Same as above.

File: include/unifex/spawn_future.hpp [hunk 4]
```diff
@@ -0,0 +1,1029 @@
...
+// Spawning a future creates a race condition between the spawned operation
+// completing and the future trying to observe the result; we resolve this race
+// by moving the operation through a state machine that reflects what work has
+// been done and who's waiting for whom.
+//
+// Successfully-spawned operations start in the init state; from there, the
+// simplest path is for the operation to complete normally before the future's
+// continuation consumes the result, in which case the operation moves from init
+// to one of value, error, or done (depending on the result) and then the
+// future deletes the spawned operation as a side effect of completing.
```
[the remark this answers is not in the document]
Comment (ispeters) [reply include/unifex/spawn_future.hpp:52]: Yes; should I change something to make that more clear?

File: include/unifex/v2/async_scope.hpp [hunk 5]
```diff
@@ -350,6 +351,27 @@ struct _nest_sender<Sender>::type final {
...
+      // lhs has no scope so if rhs *does*, we need to move-construct
+      sender_.construct(std::move(rhs).sender_.get());
+      rhs.sender_.destruct();
+    } else {
+      // no scope on either side
+    }
+
+    scope_ = std::move(rhs).scope_;
+
+    return *this;
```
Comment (ispeters) [inline include/unifex/v2/async_scope.hpp:372]:
```suggestion
    if (!scope_ && !rhs.scope_) {
      // nothing to do
      return *this;
    }
    
    UNIFEX_ASSERT(scope_ || rhs.scope_);
    
    if (scope_) {
      // lhs has a scope so we need to destroy its sender
      sender_.destruct();
    }
    
    if (rhs.scope_) {
      // rhs has a scope so move its sender to lhs
      sender_.construct(std::move(rhs).sender_.get());
      rhs.sender_.destruct();
    }

    scope_ = std::move(rhs).scope_;

    return *this;
```

File: test/spawn_future_test.cpp [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+TEST(spawn_future_test, spawn_future_of_just_and_v1_scope_compiles) {
+  unifex::v1::async_scope scope;
+
+  bool didExecute = false;
+  auto future = unifex::spawn_future(
+      unifex::just(&didExecute) |
+          unifex::then([](auto* didExecute) noexcept { *didExecute = true; }),
+      scope);
+
+  unifex::sync_wait(unifex::when_all(scope.complete(), std::move(future)));
```
[the remark this answers is not in the document]
Comment (ispeters) [reply test/spawn_future_test.cpp:47]: I'm about to push a PR that will reimplement `v1::async_scope::spawn()` in terms of this PR; we have a bunch more tests for that method, plus a bunch of internal uses, with corresponding tests, that will give me enough confidence to merge this as-is.  I come back around and add more tests later.

**Pull request merged**